src/sabyenc3.cc

```diff
@@ -571,6 +577,147 @@ PyObject* decode_usenet_chunks(PyObject* self, PyObject* Py_input_list) {
+    if (!start_loc) {
+        Py_BLOCK_THREADS;
+        PyErr_SetString(PyExc_ValueError, "Invalid yEnc header");
+        retval = NULL;
+        goto finish;
+    }
+
+    // Find start of the filename
+    start_loc = my_memstr(start_loc, end_loc - cur_char, " name=", 1);
+    if (!start_loc) {
+        Py_BLOCK_THREADS;
+        PyErr_SetString(PyExc_ValueError, "Could not find yEnc filename");
+        retval = NULL;
+        goto finish;
+    }
+
+    // Extract filename
+    cur_char = start_loc;
+    for (; *cur_char != SABYENC_LF && *cur_char != SABYENC_CR && *cur_char != SABYENC_ZERO; cur_char++);
+    Py_output_filename = PyUnicode_DecodeLatin1(start_loc, cur_char - start_loc, NULL);
+
+    // Check for =ypart, so we know where to start with decoding
+    start_loc = my_memstr(cur_char, end_loc - cur_char, "=ypart ", 1);
+    if (start_loc) {
+        // Move to end of this line
+        cur_char = start_loc;
+        for (; *cur_char != SABYENC_LF && *cur_char != SABYENC_CR && *cur_char != SABYENC_ZERO; cur_char++);
+    }
+    start_loc = cur_char;
+
+    /*
+        Looking for the end, format:
+        =yend size=384000 part=41 pcrc32=084e170f
+    */
+    // Make sure we don't go past the end of the buffer
+    if (end_loc - MAX_TAIL_BYTES > cur_char) {
+        cur_char = end_loc - MAX_TAIL_BYTES;
+    }
+    cur_char = my_memstr(cur_char, end_loc - cur_char, "\r\n=yend", 0);
+    if (!cur_char) {
+        Py_BLOCK_THREADS;
+        PyErr_SetString(PyExc_ValueError, "Invalid yEnc footer");
+        retval = NULL;
+        goto finish;
+    }
+    yenc_data_length = cur_char - start_loc;
+
+    // Try to find the crc32 of the part (skip "\r\n=yend")
+    cur_char += 7;
+    crc_pos = my_memstr(cur_char, end_loc - cur_char, " pcrc32=", 1);
+
+    // Sometimes only crc32 is used
+    if (!crc_pos) {
+        crc_pos = my_memstr(cur_char, end_loc - cur_char, " crc32=", 1);
+    }
+
+    // Parse CRC32
+    if (crc_pos && (end_loc - crc_pos) >= 8) {
+        crc_yenc = strtoul(crc_pos, NULL, 16);
+    } else {
+        // CRC32 not found - article is invalid
+        Py_BLOCK_THREADS;
+        PyErr_SetString(PyExc_ValueError, "Invalid CRC in footer");
+        retval = NULL;
+        goto finish;
+    }
+
+    // send to decoder
+    YencDecoderState state = YDEC_STATE_CRLF;
+    output_len = do_decode(1, (unsigned char*) start_loc, (unsigned char*) dest_loc, yenc_data_length, &state);
+    crc = do_crc32(dest_loc, output_len, crc);
+
+    // Return GIL to perform Python modifications
+    Py_END_ALLOW_THREADS;
+
+    // Terminate buffer and adjust the Python-size of the bytes-object
+    resize_pybytes(Py_bytes_obj, output_len);
+
+    // Is there a valid CRC?
+    if (crc != crc_yenc) {
+        Py_output_crc = Py_None;
+        Py_INCREF(Py_output_crc);
+    } else {
+        Py_output_crc = PyLong_FromUnsignedLong(crc);
+    }
+
+    // Build output
+    retval = Py_BuildValue("(S, O)", Py_output_filename, Py_output_crc);
```

Comment:
I think this leaks, O increases the reference count on Py_output_crc https://docs.python.org/3/c-api/arg.html#c.Py_BuildValue recomment checking with Py_REFCNT

```suggestion
    retval = Py_BuildValue("(S, N)", Py_output_filename, Py_output_crc);
```

Alternatively add a Py_XDECREF in finish.
Hangover from adding the CRC I think but same issue on the chunks approach.

Reply:
You are right it seems!